File: java/src/main/java/br/com/emendes/problems/WordSearch.java
package br.com.emendes.problems;

/**
 * Dado um grid board de m x n caracteres e uma string word, retorne true se word exists no grid.<br><br>
 * <p>
 * A palavra pode ser construída a partir de letras de células adjacentes sequencialmente,
 * onde as células adjacentes são vizinhas horizontal ou verticalmente.
 * A mesma célula de letra não pode ser usada mais de uma vez.<br><br>
 * <p>
 * Restrições:<br>
 * m == board.length<br>
 * n = board[i].length<br>
 * 1 <= m, n <= 6<br>
 * 1 <= word.length <= 15<br>
 * board e word consistem de letras inglesas maiúsculas e minúsculas.
 */
public class WordSearch {

  public boolean exist(char[][] board, String word) {
    if (board.length * board[0].length < word.length()) return false;

    char[] wordArray = word.toCharArray();
    if (!hasCharacters(board, wordArray)) return false;

    for (int i = 0; i < board.length; i++) {
      for (int j = 0; j < board[0].length; j++) {
        if (checkPath(i, j, board, 0, wordArray)) return true;
      }
    }

    return false;
  }

  private boolean hasCharacters(char[][] board, char[] wordArray) {
    int[] counterCharacter = new int[128];

    for (char[] chars : board) {
      for (char character : chars) {
        counterCharacter[character]++;
      }
    }
    for (char character : wordArray) {
      counterCharacter[character]--;
      if (counterCharacter[character] < 0) return false;
    }
    return true;
  }

  private boolean checkPath(int i, int j, char[][] board, int indexCharacter, char[] word) {
    if (indexCharacter == word.length) return true;
    if (i < 0 || i >= board.length || j < 0 || j >= board[0].length) return false;
    if (board[i][j] != word[indexCharacter] || board[i][j] == '.') return false;

    char cache = board[i][j];
    board[i][j] = '.';
    boolean isGoodPath = checkPath(i, j + 1, board, indexCharacter + 1, word) ||
        checkPath(i + 1, j, board, indexCharacter + 1, word) ||
        checkPath(i, j - 1, board, indexCharacter + 1, word) ||
        checkPath(i - 1, j, board, indexCharacter + 1, word);

    board[i][j] = cache;
    return isGoodPath;
  }

}

File: javascript/problems/WordSearch.js
/**
 * Dado um grid board de m x n caracteres e uma string word,
 * retorne true se word existe no grid.
 *
 * A palavra pode ser construída por letras sequênciais de células adjacentes,
 * onde células adjacentes são vizinhos horizontais ou verticais.
 * Uma mesma célula não pode ser usado mais de uma vez.
 *
 * Restrições:
 *
 * m == board.length
 *
 * n = board[i].length
 *
 * 1 <= m, n <= 6
 *
 * 1 <= word.length <= 15
 *
 * board e word consistem de letras minúsculas.
 *
 * Follow-up: Você poderia usar a poda de pesquisa para tornar sua solução mais rápida com um quadro maior?
 */
const exist = function (board, word) {
  const used = new Array(board.length);
  for (let i = 0; i < board.length; i++) {
    used[i] = new Array(board[i].length).fill(false);
  }
  for (let i = 0; i < board.length; i++) {
    for (let j = 0; j < board[0].length; j++)
      if (existWord(board, word, 0, i, j, used)) return true;
  }
  return false;
};

const existWord = (board, word, wi, bi, bj, used) => {
  if (wi === word.length) return true;
  if (bi < 0 || bi >= board.length || bj < 0 || bj >= board[0].length)
    return false;
  if (used[bi][bj] || word[wi] !== board[bi][bj]) return false;

  used[bi][bj] = true;
  const result =
    existWord(board, word, wi + 1, bi + 1, bj, used) ||
    existWord(board, word, wi + 1, bi - 1, bj, used) ||
    existWord(board, word, wi + 1, bi, bj + 1, used) ||
    existWord(board, word, wi + 1, bi, bj - 1, used);
  used[bi][bj] = false;
  return result;
};
